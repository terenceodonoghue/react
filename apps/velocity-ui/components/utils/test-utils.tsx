/* eslint-disable import/no-extraneous-dependencies */
import React, { FunctionComponent, ReactElement } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import ThemeProvider, {
  ThemeProviderProps,
} from '~/components/core/ThemeProvider';

const Provider: FunctionComponent<ThemeProviderProps> = ({ children }) => (
  <ThemeProvider>{children}</ThemeProvider>
);

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
): RenderResult => render(ui, { wrapper: Provider, ...options });

export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';

export { customRender as render };
