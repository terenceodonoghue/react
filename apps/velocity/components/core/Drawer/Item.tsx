import { useTheme } from '@emotion/react';
import { rgba, transitions } from 'polished';
import {
  forwardRef,
  JSXElementConstructor,
  LiHTMLAttributes,
  SVGProps,
} from 'react';

export interface ItemProps extends LiHTMLAttributes<HTMLLIElement> {
  compact?: boolean;
  icon?: JSXElementConstructor<SVGProps<SVGSVGElement>>;
  selected?: boolean;
}

const Item = forwardRef<HTMLLIElement, ItemProps>(
  (
    { children, compact = false, icon: Icon, selected = false, ...props },
    ref,
  ) => {
    const theme = useTheme();

    return (
      <li
        css={[
          ({ color, transition }) => ({
            alignItems: 'center',
            backgroundColor: selected ? rgba(color.primary, 0.1) : undefined,
            borderLeft: `solid 3px ${selected ? color.primary : 'transparent'}`,
            color: selected ? color.primary : color.neutral[500],
            cursor: 'pointer',
            display: 'flex',
            height: 56,
            justifyContent: 'flex-start',
            padding: '0 0 0 25px',
            ...transitions(['background-color'], transition.quickly),
            '&:hover': {
              backgroundColor: !selected && rgba(color.primary, 0.1),
            },
          }),
        ]}
        data-testid="drawer-item"
        ref={ref}
        {...props}
      >
        {Icon && (
          <Icon
            color={theme.color.neutral[400]}
            css={{
              flexShrink: 0,
              height: 24,
              width: 24,
            }}
            data-testid="drawer-item-icon"
          />
        )}
        <span
          css={[
            ({ transition }) => ({
              marginLeft: 16,
              opacity: compact ? 0 : 1,
              ...transitions(['opacity'], transition.quickly),
            }),
          ]}
          data-testid="drawer-item-text"
        >
          {children}
        </span>
      </li>
    );
  },
);

export default Item;
