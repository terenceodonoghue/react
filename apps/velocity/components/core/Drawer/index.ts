import Drawer from './Drawer';

export { default as DrawerItem } from './Item';
export default Drawer;
