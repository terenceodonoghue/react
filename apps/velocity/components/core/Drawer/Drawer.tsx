import { Avatar } from '@terenceodonoghue/react-components/velocity';
import { FunctionComponent, HTMLAttributes } from 'react';
import { transitions as transition } from 'polished';
import { useMediaQuery } from 'react-responsive';
import { animated, config, useSpring } from 'react-spring';
import { DESKTOP } from '../../utils/breakpoints';
import Item, { ItemProps } from './Item';

interface DrawerProps extends HTMLAttributes<HTMLDivElement> {
  anchor?: 'left' | 'right';
  avatar: string;
  name: string;
  open?: boolean;
}

interface DrawerComposition {
  Item: FunctionComponent<ItemProps>;
}

const Drawer: FunctionComponent<DrawerProps> & DrawerComposition = ({
  anchor = 'left',
  avatar,
  children,
  name,
  open = false,
  ...props
}) => {
  const isTabletOrMobile = useMediaQuery({
    maxWidth: DESKTOP - 1,
  });

  const closedWidth = isTabletOrMobile ? 0 : 80;

  const { width } = useSpring({
    config: config.stiff,
    width: open ? 215 : closedWidth,
  });

  return (
    <animated.div
      css={[
        ({ palette, zIndex }) => ({
          backgroundColor: palette.neutral[50],
          bottom: 0,
          overflowX: 'hidden',
          padding: '20px 0',
          position: 'fixed',
          top: 79,
          zIndex: zIndex.drawer,
        }),
        anchor === 'left' ? { left: 0 } : { right: 0 },
      ]}
      data-testid="drawer"
      style={{ width }}
      {...props}
    >
      <div
        css={[
          ({ transitions }) => ({
            display: 'flex',
            alignItems: 'center',
            margin: '27px 0 34px',
            opacity: open ? 1 : 0,
            paddingLeft: 28,
            ...transition(
              'opacity',
              `${transitions.duration.standard}ms ${transitions.easing.easeInOut}`,
            ),
          }),
        ]}
      >
        <Avatar
          alt={name}
          css={{
            marginRight: 8,
          }}
          size={32}
          src={avatar}
          variant="square"
        />
        <div>
          <span>Welcome</span>
          <span
            css={({ palette }) => ({
              color: palette.neutral[400],
              display: 'block',
              whiteSpace: 'nowrap',
            })}
          >
            {name}
          </span>
        </div>
      </div>
      <nav>
        <ul
          css={{
            listStyle: 'none',
            margin: 0,
            padding: 0,
          }}
        >
          {children}
        </ul>
      </nav>
    </animated.div>
  );
};

Drawer.Item = Item;

export default Drawer;
