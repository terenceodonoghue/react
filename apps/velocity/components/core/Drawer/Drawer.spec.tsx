import React from 'react';
import faker from 'faker';
import { render } from 'test-utils';
import Drawer from './Drawer';

describe('Drawer', () => {
  it('is closed by default', () => {
    const avatar = faker.image.avatar();
    const name = faker.name.findName();
    const { container, getByTestId } = render(
      <Drawer avatar={avatar} name={name} />,
    );

    expect(getByTestId('drawer')).toHaveStyle({ width: '80px' });
    expect(container).toMatchSnapshot();
  });

  it('expands when opened', () => {
    const avatar = faker.image.avatar();
    const name = faker.name.findName();
    const { container, getByTestId } = render(
      <Drawer avatar={avatar} name={name} open />,
    );

    expect(getByTestId('drawer')).toHaveStyle({ width: '215px' });
    expect(container).toMatchSnapshot();
  });

  it('anchors to the left', () => {
    const avatar = faker.image.avatar();
    const name = faker.name.findName();
    const { container, getByTestId } = render(
      <Drawer anchor="left" avatar={avatar} name={name} />,
    );

    expect(getByTestId('drawer')).toHaveStyle({ left: 0 });
    expect(container).toMatchSnapshot();
  });

  it('anchors to the right', () => {
    const avatar = faker.image.avatar();
    const name = faker.name.findName();
    const { container, getByTestId } = render(
      <Drawer anchor="right" avatar={avatar} name={name} />,
    );

    expect(getByTestId('drawer')).toHaveStyle({ right: 0 });
    expect(container).toMatchSnapshot();
  });
});
