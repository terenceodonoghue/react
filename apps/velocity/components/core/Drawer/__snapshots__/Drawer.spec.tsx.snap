// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer anchors to the left 1`] = `
.emotion-0 {
  background-color: #fff;
  bottom: 0;
  overflow-x: hidden;
  padding: 20px 0;
  position: fixed;
  top: 79px;
  z-index: 1200;
  left: 0;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 27px 0 34px;
  opacity: 0;
  padding-left: 28px;
  -webkit-transition: opacity,300ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: opacity,300ms cubic-bezier(0.4, 0, 0.2, 1);
}

.emotion-2 {
  border-radius: 6px;
  height: 32px;
  width: 32px;
  margin-right: 8px;
}

.emotion-3 {
  color: #bfc5d2;
  display: block;
  white-space: nowrap;
}

.emotion-4 {
  list-style: none;
  margin: 0;
  padding: 0;
}

<div>
  <div
    class="emotion-0"
    data-testid="drawer"
    style="width: 80px;"
  >
    <div
      class="emotion-1"
    >
      <img
        alt="Terri Paucek"
        class="emotion-2"
        data-testid="avatar"
        src="https://cdn.fakercloud.com/avatars/iqonicd_128.jpg"
      />
      <div>
        <span>
          Welcome
        </span>
        <span
          class="emotion-3"
        >
          Terri Paucek
        </span>
      </div>
    </div>
    <nav>
      <ul
        class="emotion-4"
      />
    </nav>
  </div>
</div>
`;

exports[`Drawer anchors to the right 1`] = `
.emotion-0 {
  background-color: #fff;
  bottom: 0;
  overflow-x: hidden;
  padding: 20px 0;
  position: fixed;
  top: 79px;
  z-index: 1200;
  right: 0;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 27px 0 34px;
  opacity: 0;
  padding-left: 28px;
  -webkit-transition: opacity,300ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: opacity,300ms cubic-bezier(0.4, 0, 0.2, 1);
}

.emotion-2 {
  border-radius: 6px;
  height: 32px;
  width: 32px;
  margin-right: 8px;
}

.emotion-3 {
  color: #bfc5d2;
  display: block;
  white-space: nowrap;
}

.emotion-4 {
  list-style: none;
  margin: 0;
  padding: 0;
}

<div>
  <div
    class="emotion-0"
    data-testid="drawer"
    style="width: 80px;"
  >
    <div
      class="emotion-1"
    >
      <img
        alt="Charlie Jacobs"
        class="emotion-2"
        data-testid="avatar"
        src="https://cdn.fakercloud.com/avatars/m_kalibry_128.jpg"
      />
      <div>
        <span>
          Welcome
        </span>
        <span
          class="emotion-3"
        >
          Charlie Jacobs
        </span>
      </div>
    </div>
    <nav>
      <ul
        class="emotion-4"
      />
    </nav>
  </div>
</div>
`;

exports[`Drawer expands when opened 1`] = `
.emotion-0 {
  background-color: #fff;
  bottom: 0;
  overflow-x: hidden;
  padding: 20px 0;
  position: fixed;
  top: 79px;
  z-index: 1200;
  left: 0;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 27px 0 34px;
  opacity: 1;
  padding-left: 28px;
  -webkit-transition: opacity,300ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: opacity,300ms cubic-bezier(0.4, 0, 0.2, 1);
}

.emotion-2 {
  border-radius: 6px;
  height: 32px;
  width: 32px;
  margin-right: 8px;
}

.emotion-3 {
  color: #bfc5d2;
  display: block;
  white-space: nowrap;
}

.emotion-4 {
  list-style: none;
  margin: 0;
  padding: 0;
}

<div>
  <div
    class="emotion-0"
    data-testid="drawer"
    style="width: 215px;"
  >
    <div
      class="emotion-1"
    >
      <img
        alt="Beulah Kulas"
        class="emotion-2"
        data-testid="avatar"
        src="https://cdn.fakercloud.com/avatars/nerdgr8_128.jpg"
      />
      <div>
        <span>
          Welcome
        </span>
        <span
          class="emotion-3"
        >
          Beulah Kulas
        </span>
      </div>
    </div>
    <nav>
      <ul
        class="emotion-4"
      />
    </nav>
  </div>
</div>
`;

exports[`Drawer is closed by default 1`] = `
.emotion-0 {
  background-color: #fff;
  bottom: 0;
  overflow-x: hidden;
  padding: 20px 0;
  position: fixed;
  top: 79px;
  z-index: 1200;
  left: 0;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 27px 0 34px;
  opacity: 0;
  padding-left: 28px;
  -webkit-transition: opacity,300ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: opacity,300ms cubic-bezier(0.4, 0, 0.2, 1);
}

.emotion-2 {
  border-radius: 6px;
  height: 32px;
  width: 32px;
  margin-right: 8px;
}

.emotion-3 {
  color: #bfc5d2;
  display: block;
  white-space: nowrap;
}

.emotion-4 {
  list-style: none;
  margin: 0;
  padding: 0;
}

<div>
  <div
    class="emotion-0"
    data-testid="drawer"
    style="width: 80px;"
  >
    <div
      class="emotion-1"
    >
      <img
        alt="Duane Farrell"
        class="emotion-2"
        data-testid="avatar"
        src="https://cdn.fakercloud.com/avatars/nicklacke_128.jpg"
      />
      <div>
        <span>
          Welcome
        </span>
        <span
          class="emotion-3"
        >
          Duane Farrell
        </span>
      </div>
    </div>
    <nav>
      <ul
        class="emotion-4"
      />
    </nav>
  </div>
</div>
`;
